File: reborn/analysis/peaks.py
# This file is part of reborn <https://kirianlab.gitlab.io/reborn/>.
#
# reborn is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# reborn is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with reborn.  If not, see <https://www.gnu.org/licenses/>.

import numpy as np
from scipy.ndimage import measurements
from ..fortran import peaks_f
from ..utils import ensure_list
from ..detector import RadialProfiler, concat_pad_data, split_pad_data


class MultiPADPeakFinder(object):
    r"""
    A peak finder class that deals with multiple PADs.
    """

    def __init__(self, snr_threshold=6, radii=(1, 5, 10), mask=None, subtract_radial_median=False, correct_polarization=False, pad_geometry=None, beam=None):
        r"""
        Args:
            snr_threshold: Peaks must have a signal-to-noise ratio above this value
            radii: These are the radii associated with the :func:`boxsnr <reborn.analysis.peaks.boxsnr>` function.
            mask: Ignore pixels where mask == 0.
            subtract_radial_median: Boolean variable to switch on radial median subtraction
            correct_polarization: Boolean variable to switch on polarization correction
            pad_geometry: pad geometry object
            beam: beam object
        """
        pad_geometry = ensure_list(pad_geometry)

        self.pads = pad_geometry
        n_pads = len(self.pads)
        
        self.snr_threshold = snr_threshold
        self.radii = radii
        
        if mask is None:
            self.mask = [None] * n_pads
        self.mask = ensure_list(self.mask)

        self.peak_finders = [PeakFinder(snr_threshold=snr_threshold, radii=radii, mask=self.mask[i]) for i in range(len(self.pads))]

        self.subtract_radial_median = subtract_radial_median
        self.correct_polarization = correct_polarization

        if subtract_radial_median:
            qmag = np.concatenate([p.q_mags(beam=beam) for p in self.pads])
            self.radial_profiler = RadialProfiler(n_bins=100,
                                                  q_range=(np.min(qmag), np.max(qmag)),
                                                  pad_geometry=self.pads,
                                                  beam=beam)
            self.qmag = qmag
        if correct_polarization:
            self.polarization_factor = concat_pad_data([p.polarization_factors(beam=beam) for p in self.pads])


    def find_peaks(self, data, mask=None):
        r"""
        Do peak-finding on a data array with multiple PADs.

        Args:
            data: The data to perform peak-finding on.
            mask: A mask, if desired.  By default, the mask used on creation of a PeakFinder instance will be used.
                  This defaults to ones everywhere.

        Returns:
            A list of centroids of the peaks.
        """

        # Put the data PADs into a 1D array
        data = concat_pad_data(data)

        # Divide out polarization
        if self.correct_polarization:
            data /= self.polarization_factor

        # Calculate and subtract the radial median
        if self.subtract_radial_median:
            radial_med = self.radial_profiler.get_median_profile(data)

            data -= np.interp(self.qmag, self.radial_profiler.bin_centers, radial_med)

        # Put the data back into PADs 
        data = split_pad_data(self.pads, data)

        # Find the peaks for each PAD
        peaks_centroids = []
        for i in range(len(self.pads)):
            peaks_centroids.append(self.peak_finders[i].find_peaks(data=data[i]))

        return peaks_centroids


class PeakFinder(object):
    r"""
    A crude peak finder.  It works by firstly running a signal-to-noise filter over an entire image, and then searching
    for connected regions with SNR above some threshold.  It is not fully developed.  For example, it does not yet
    check for the minimum distance between peaks.
    """

    mask = None  #: Ignore pixels where mask == 0.
    snr_threshold = None  # : Peaks must have a signal-to-noise ratio above this value.
    radii = None  # : These are the radii associated with the :func:`boxsnr <reborn.analysis.peaks.boxsnr>` function.
    snr = None  # : The SNR array from the most recent call to the find_peaks method.
    signal = None  # : The signal array from the most recent call to the find_peaks method.
    labels = None  # : Labeled regions of peak candidates
    n_labels = 0  # : Number of peak candidates
    centroids = None  # : Centroids of each peak candidate
    beam = None  # :

    def __init__(self, snr_threshold=6, radii=(3, 5, 10), mask=None, max_iterations=3, beam=None):
        r"""
        Args:
            snr_threshold: Peaks must have a signal-to-noise ratio above this value
            radii: These are the radii associated with the :func:`boxsnr <reborn.analysis.peaks.boxsnr>` function.
            mask: Ignore pixels where mask == 0.
        """
        self.snr_threshold = snr_threshold
        self.radii = radii
        self.mask = mask
        self.max_iterations = max_iterations

    def find_peaks(self, data, mask=None):
        r"""
        Do peak finding on a data array.

        Args:
            data: The data to perform peak finding on.
            mask: A mask, if desired.  By default, the mask used on creation of a PeakFinder instance will be used.
                  This defaults to ones everywhere.

        Returns:

        """
        if mask is None:
            if self.mask is None:
                self.mask = np.ones_like(data)
            mask = self.mask
        mask_a = mask.copy()
        for i in range(self.max_iterations):
            snr, signal = boxsnr(data, mask, mask_a, self.radii[0], self.radii[1], self.radii[2])
            ab = snr > self.snr_threshold
            if np.sum(ab) > 0:
                mask_a[ab] = 0
            else:
                break
        self.snr, self.signal = boxsnr(data, mask, mask, self.radii[0], self.radii[1], self.radii[2])
        self.labels, self.n_labels = measurements.label(self.snr > self.snr_threshold)
        #print('self.n_labels', self.n_labels)
        if self.n_labels > 0:
            sig = self.signal.copy()
            sig[sig < 0] = 0
            cent = measurements.center_of_mass(sig, self.labels, np.arange(1, self.n_labels+1))
            cent = np.array(cent)
            if len(cent.shape) == 1:
                cent = np.expand_dims(cent, axis=0)
            cent = cent[:, ::-1].copy()
            #print('cent', cent)
            self.centroids = cent
        else:
            self.centroids = None

        return self.centroids


def boxsnr(dat, mask_in, mask_out, n_in, n_cent, n_out):

    r"""
    Transform an 2D image into a map of local signal-to-noise ratio by the following equivalent steps:

    (1) For every pixel in the input data, do a local signal integration within a square region of size
        :math:`n_\text{in}*2+1`.  Pixels masked by `mask_in` will be ignored.  Masked pixels are indicated by the value
        zero, while unmasked pixels are indicated by the value one.

    (2) Estimate background via a local integration within a square annulus of outer size 
        :math:`2 n_\text{out} + 1` and inner size :math:`2 n_\text{cent} - 1`.  Pixels within `mask_out` will be
        ignored.

    (3) From every pixel in the local signal integration square, subtract the average background value from step (2).

    (4) Compute the standard deviation :math:`\sigma` in the square annulus.  Pixels within `mask_out` will be ignored.

    (5) Divide the locally-integrated signal-minus-background by the standard error.  The standard error is 
        equal to :math:`\sigma/\sqrt{M}` where :math:`M` is the number of unmasked pixels in the locally-integratied
        signal region, and :math:`\sigma` comes from step (4).

    The use of two distinct masks allows for multi-pass SNR computations in which the results of the first pass
    may be used to exclude high-SNR regions from contributing to error estimates in the annulus.  See
    :func:`snr_mask <reborn.analysis.peaks.snr_mask>` if you want to generate a mask this way.

    Note:
        This routine will attempt to use openmp to parallelize the computations.  It is affected by the environment
        variable `OMP_NUM_THREADS`.  You can check how many cores are used by openmp by running the following:

        .. code-block::

            import reborn.fortran; reborn.fortran.omp_test_f.omp_test()

    Arguments:
        dat (|ndarray|): The image to analyze.
        mask_in (|ndarray|): The mask for the square central integration region.
        mask_out (|ndarray|): The mask for the square annulus integration region.
        n_in (int): Size of the central integration region; integrate from :math:`(-n_{in}, n_{in})`, inclusively.
        n_cent (int): Define the annulus integration region; we ignore the box from (-n_cent, n_cent), inclusively.
        n_out (int): Define the annulus integration region; we include the box from (-n_out, n_out), inclusively.

    Returns:
        (tuple):

        **snr** (|ndarray|): The SNR array.

        **signal** (|ndarray|): The signal array.
    """
    float_t = np.float64
    snr = np.asfortranarray(np.ones(dat.shape, dtype=float_t))
    signal = np.asfortranarray(np.ones(dat.shape, dtype=float_t))
    d = np.asfortranarray(dat.astype(float_t))
    m = np.asfortranarray(mask_in.astype(float_t))
    m2 = np.asfortranarray(mask_out.astype(float_t))
    peaks_f.boxsnr(d, m, m2, snr, signal, n_in, n_cent, n_out)
    return snr, signal


def snr_filter(dat, mask, nin, ncent, nout, threshold=6, mask_negative=True, max_iterations=1):
    r"""
    Transform an image into a map of local signal-to-noise ratio.

    Arguments:
        dat (numpy array) : Input data to calculate SNR from.
        mask (numpy array) : Mask indicating bad pixels (zero means bad, one means ok)
        nin (int) : See boxsnr function.
        ncent (int) : See boxsnr function.
        nout (int) : See boxsnr function.
        threshold (float) : Reject pixels above this SNR.
        mask_negative (bool) : Also reject pixels below the negative of the SNR threshold (default: True).
        max_iterations (int) : The maxumum number of iterations (note: the loop exits if the mask stops changing).

    Returns:
        numpy array : The SNR map
    """
    if isinstance(dat, list):
        return [snr_filter(d, m, nin, ncent, nout, threshold=threshold, mask_negative=mask_negative,
                           max_iterations=max_iterations) for (d, m) in zip(dat, mask)]
    mask_a = mask.copy()
    prev = 0
    for i in range(max_iterations):
        a, _ = boxsnr(dat, mask, mask_a, nin, ncent, nout)
        if mask_negative:
            a = np.abs(a)
        ab = a > threshold
        above = np.sum(ab)
        mask_a[ab] = 0
        if above == prev:
            break
        prev = above
    return a


def snr_mask(dat, mask, nin, ncent, nout, threshold=6, mask_negative=True, max_iterations=3):
    r"""
    Mask out pixels above some chosen SNR threshold.  The image is converted to a map of SNR using boxsnr.  Additional
    iterations follow, in which pixels above threshold in the previous run are also masked in the annulus.
    This iterative procedure helps avoid contributions of peak signals to the Noise calculation.

    Arguments:
        dat (numpy array) : Input data to calculate SNR from.
        mask (numpy array) : Mask indicating bad pixels (zero means bad, one means ok)
        nin (int) : See boxsnr function.
        ncent (int) : See boxsnr function.
        nout (int) : See boxsnr function.
        threshold (float) : Reject pixels above this SNR.
        mask_negative (bool) : Also reject pixels below the negative of the SNR threshold (default: True).
        max_iterations (int) : The maxumum number of iterations (note: the loop exits if the mask stops changing).

    Returns:
        numpy array : The mask with pixels above the SNR threshold
    """
    if isinstance(dat, list):
        return [snr_filter(d, m, nin, ncent, nout, threshold=threshold, mask_negative=mask_negative,
                           max_iterations=max_iterations) for (d, m) in zip(dat, mask)]
    mask_a = mask.copy()
    prev = 0
    for i in range(max_iterations):
        a, _ = boxsnr(dat, mask, mask_a, nin, ncent, nout)
        if mask_negative:
            a = np.abs(a)
        ab = a > threshold
        above = np.sum(ab)
        mask_a[ab] = 0
        if above == prev:
            break
        prev = above
    return mask_a

File: reborn/fortran/peaks.f90
! This file is part of reborn <https://kirianlab.gitlab.io/reborn/>.
!
! reborn is free software: you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation, either version 3 of the License, or
! (at your option) any later version.
!
! reborn is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License
! along with reborn.  If not, see <https://www.gnu.org/licenses/>.

subroutine boxsnr(dat,mask,mask2,snr,signal,npx,npy,n_inner,n_center,n_outer)
!
! This routine computes a local signal-to-noise ratio by the following equivalent steps:
!
!   (1) For every pixel in the input data, do a local signal integration within a square region of size n_inner*2+1.
!   (2) Estimate background via a local integration within a square annulus of outer size n_outer*2 + 1 and inner size
!       n_center*2 - 1.
!   (3) From every pixel in the local signal integration square, subtract the average background value from step (2).
!   (4) Compute the standard deviation (sigma) in the square annulus.
!   (5) Divide the locally-integrated signal-minus-background by the standard error.  This standard error is equal to
!       sigma*sqrt(M) where M is the number of pixels in the locally integratied signal region, and sigma comes from
!       step (4).
!
!   Note: There are two masks that are used: one mask is for the local signal integration, and the other mask is used
!         for the local background and error annulus calculations.  The value M above, for example, is computed
!         according to the number of valid pixels, which are indicated by mask == 1.  The use of two masks allows for
!         multi-pass SNR computations in which the results of the first pass may be used to exclude high-SNR regions
!         from contributing to error estimates in the annulus.
!
!   Note: This routine will attempt to use openmp to parallelize the computations.  It is affected by the environment
!         variable OMP_NUM_THREADS
!
!                        ----------------------------------------
!                        |                                      |
!                        |Compute NOISE & BACKGROUND in annulus |
!                        |                                      |
!                        |      |------------------------|      |
!                        |      |  ignore this region    |      |
!                        |      |    |--------------|    |      |
!                        |      |    | Compute      |    |      |
!                        |      |    | integrated   |    |      |
!                        |      |    | SIGNAL minus |    |      |
!                        |      |    | BACKGROUND   |    |      |
!                        |      |    | here         |    |      |
!                        |      |    |--------------|    |      |
!                        |      |        n_inner         |      |
!                        |      |------------------------|      |
!                        |               n_center               |
!                        |                                      |
!                        |                                      |
!                        ----------------------------------------
!                                        n_outer
!
! dat : The 2D array to run the SNR filter over.
! mask : Ignore signal pixels where mask == 0 (the signal is multiplied by this mask).
! mask2 : Ignore background/noise pixels where mask2 == 0 (the background is multiplied by this mask).
! snr : The output SNR
! signal : The output background subtracted signal
! n_inner, n_center, n_outer : define the integration regions: see figure above

    real(kind=8), intent(in) :: dat(npx,npy), mask(npx,npy), mask2(npx,npy)
    real(kind=8), intent(inout) :: snr(npx,npy), signal(npx,npy)
    integer(kind=4), intent(in) :: n_inner, n_center, n_outer
    real(kind=8) :: cumix(0:npx,npy), cummix(0:npx,npy), &
                    cumox(0:npx,npy), cum2ox(0:npx,npy), cummox(0:npx,npy), &
                    cumiy(npx,0:npy), cummiy(npx,0:npy), &
                    cumcy(npx,0:npy), cum2cy(npx,0:npy), cummcy(npx,0:npy), &
                    cumoy(npx,0:npy), cum2oy(npx,0:npy), cummoy(npx,0:npy), &
                    sqix(npx,npy), sqmix(npx,npy), &
                    sqcx(npx,npy), sq2cx(npx,npy), sqmcx(npx,npy), &
                    sqox(npx,npy), sq2ox(npx,npy), sqmox(npx,npy)
    real(kind=8) :: small
    integer(kind=4) :: ix,iy,mn,mx,npx,npy

    small=1.0e-15_8

    !$OMP parallel default(None) private(ix,iy,mn,mx) &
    !$OMP shared(dat,mask,mask2,cumix,cum2ix,cummix,cumox,cum2ox,cummox,cumiy,sqix,cum2iy,sq2ix,cummiy,sqmix,cumcy, &
    !$OMP sqcx,cum2cy,sq2cx,cummcy,sqmcx,cumoy,sqox,cum2oy,sq2ox,cummoy,sqmox,npx,npy,n_inner,n_center,n_outer, &
    !$OMP small,snr,signal)

    !$OMP do schedule(static)
    do iy=1,npy
        cumix(0,iy)=0.0_8  ! cumulative data inner x
        cummix(0,iy)=0.0_8 ! cumulative mask inner x

        cumox(0,iy)=0.0_8  ! outer
        cum2ox(0,iy)=0.0_8
        cummox(0,iy)=0.0_8
    enddo
    !$OMP enddo nowait

    !$OMP do schedule(static)
    do ix=1,npx
        cumiy(ix,0)=0.0_8  ! inner
        cummiy(ix,0)=0.0_8

        cumcy(ix,0)=0.0_8  ! center
        cum2cy(ix,0)=0.0_8
        cummcy(ix,0)=0.0_8

        cumoy(ix,0)=0.0_8  ! outer
        cum2oy(ix,0)=0.0_8
        cummoy(ix,0)=0.0_8
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do iy=1,npy
        do ix=1,npx  ! cumulative sums
            cumix(ix,iy)=cumix(ix-1,iy)+dat(ix,iy)*mask(ix,iy)
            cummix(ix,iy)=cummix(ix-1,iy)+mask(ix,iy)

            cumox(ix,iy)=cumox(ix-1,iy)+dat(ix,iy)*mask2(ix,iy)
            cum2ox(ix,iy)=cum2ox(ix-1,iy)+(dat(ix,iy)*mask2(ix,iy))**2
            cummox(ix,iy)=cummox(ix-1,iy)+mask2(ix,iy)
        enddo
    enddo
    !$OMP enddo


    !$OMP do schedule(static)
    do ix=1,npx  ! windowed sums on one axis
        do iy=1,npy
            mn=min(npx,ix+n_inner)
            mx=max(0,ix-n_inner-1)
            sqix(ix,iy) =cumix(mn,iy)-cumix(mx,iy)
            sqmix(ix,iy)=cummix(mn,iy)-cummix(mx,iy)

            mn=min(npx,ix+n_center)
            mx=max(0,ix-n_center-1)
            sqcx(ix,iy) =cumox(mn,iy)-cumox(mx,iy)
            sq2cx(ix,iy)=cum2ox(mn,iy)-cum2ox(mx,iy)
            sqmcx(ix,iy)=cummox(mn,iy)-cummox(mx,iy)

            mn=min(npx,ix+n_outer)
            mx=max(0,ix-n_outer-1)
            sqox(ix,iy) =cumox(mn,iy)-cumox(mx,iy)
            sq2ox(ix,iy)=cum2ox(mn,iy)-cum2ox(mx,iy)
            sqmox(ix,iy)=cummox(mn,iy)-cummox(mx,iy)
        enddo
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do ix=1,npx
        do iy=1,npy
            cumiy(ix,iy) =cumiy(ix,iy-1) +sqix(ix,iy)
            cummiy(ix,iy)=cummiy(ix,iy-1)+sqmix(ix,iy)

            cumcy(ix,iy) =cumcy(ix,iy-1) +sqcx(ix,iy)
            cum2cy(ix,iy)=cum2cy(ix,iy-1)+sq2cx(ix,iy)
            cummcy(ix,iy)=cummcy(ix,iy-1)+sqmcx(ix,iy)

            cumoy(ix,iy) =cumoy(ix,iy-1) +sqox(ix,iy)
            cum2oy(ix,iy)=cum2oy(ix,iy-1)+sq2ox(ix,iy)
            cummoy(ix,iy)=cummoy(ix,iy-1)+sqmox(ix,iy)
        enddo
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do iy=1,npy
        do ix=1,npx
            mn=min(npy,iy+n_inner)
            mx=max(0,iy-n_inner-1)
            sqix(ix,iy) =cumiy (ix,mn)-cumiy(ix,mx)
            sqmix(ix,iy)=cummiy(ix,mn)-cummiy(ix,mx)

            mn=min(npy,iy+n_center)
            mx=max(0,iy-n_center-1)
            sqcx(ix,iy) =cumcy (ix,mn)-cumcy(ix,mx)
            sq2cx(ix,iy) =cum2cy (ix,mn)-cum2cy(ix,mx)
            sqmcx(ix,iy)=cummcy(ix,mn)-cummcy(ix,mx)

            mn=min(npy,iy+n_outer)
            mx=max(0,iy-n_outer-1)
            sqox(ix,iy) =cumoy (ix,mn)-cumoy(ix,mx)
            sq2ox(ix,iy)=cum2oy(ix,mn)-cum2oy(ix,mx)
            sqmox(ix,iy)=cummoy(ix,mn)-cummoy(ix,mx)
        enddo
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do iy=1,npy
        do ix=1,npx
            sqox(ix,iy) = sqox(ix,iy) - sqcx(ix,iy)
            sq2ox(ix,iy) = sq2ox(ix,iy) - sq2cx(ix,iy)
            sqmox(ix,iy) = sqmox(ix,iy) - sqmcx(ix,iy) + small ! avoid divide by zero
            sqmix(ix,iy) = sqmix(ix,iy) + small         ! avoid divide by zero
            signal(ix,iy) = sqix(ix,iy) - sqox(ix,iy)*sqmix(ix,iy)/sqmox(ix,iy)
            snr(ix,iy) = signal(ix,iy)/(sqrt(sqmix(ix,iy))*(sqrt(sq2ox(ix,iy)/sqmox(ix,iy) - (sqox(ix,iy)/sqmox(ix,iy))**2)+small))
        enddo
    enddo
    !$OMP enddo
    !$OMP end parallel

end subroutine boxsnr


subroutine boxconv(dat,datconv,n, npx, npy)

!
! dat : The 2D array to convolve
! datconv : The output convolved array
! n : The width of the convolution kernel
!

    real(kind=8), intent(in) :: dat(npx,npy)
    real(kind=8), intent(inout) :: datconv(npx,npy)
    integer(kind=4), intent(in) :: n
    real(kind=8) :: cumix(0:npx,npy), cumiy(npx,0:npy)
    integer(kind=4) :: ix,iy,mn,mx,npx,npy

    !$OMP parallel default(None) private(ix,iy,mn,mx) &
    !$OMP shared(dat,cumix,cumiy,datconv,sqcx,cum2cy,sq2cx,npx,npy,n)

    !$OMP do schedule(static)
    do iy=1,npy
        cumix(0,iy) = 0.0_8  ! cumulative data inner x
    enddo
    !$OMP enddo nowait

    !$OMP do schedule(static)
    do ix=1,npx
        cumiy(ix,0) = 0.0_8  ! inner
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do iy=1,npy
        do ix=1,npx  ! cumulative sums
            cumix(ix,iy) = cumix(ix-1,iy)+dat(ix,iy)
        enddo
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do ix=1,npx  ! windowed sums on one axis
        do iy=1,npy
            mn=min(npx,ix+n)
            mx=max(0,ix-n-1)
            datconv(ix,iy) = cumix(mn,iy)-cumix(mx,iy)
        enddo
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do ix=1,npx
        do iy=1,npy
            cumiy(ix,iy) = cumiy(ix,iy-1) + datconv(ix,iy)
        enddo
    enddo
    !$OMP enddo

    !$OMP do schedule(static)
    do iy=1,npy
        do ix=1,npx
            mn=min(npy,iy+n)
            mx=max(0,iy-n-1)
            datconv(ix,iy) = cumiy(ix,mn)-cumiy(ix,mx)
        enddo
    enddo
    !$OMP enddo
    !$OMP end parallel

end subroutine boxconv
